docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -461,7 +465,7 @@ The proposal calls for the implementation of an infrastructure cluster template
 ##### For Control plane providers
 Similarly, a control plane provider should also create a template for use of creation for the control plane objects. For instance, the kubeadm control plane provider should introduce the notion of a KubeadmControlPlaneTemplate for use by the cluster controller to create a managed Kubeadm Control plane object.
 
-The CAPI Cluster Controller would use this template to instantiate a control plane object for the topology. The current CRD contract mandates that the control plane provider supports `replicas` in the `spec` fields. With the introduction of ClusterClass which is responsible for handling the kubernetes version too, the current contract needs to be expanded to include the support for `version` in the `spec` fields.
+The CAPI Cluster Controller would use this template to instantiate a control plane object for the topology. With the introduction of ClusterClass which is responsible for handling the kubernetes version too, the current contract needs to be expanded to include the support for `version` in the `spec` fields. The control plane templates can optinally also support `replicas` in `spec` and its value will be set to `cluster.spec.topology.controlPlane.replicas`.
```

Comment:
I think the whole paragraph has to be rephrased a bit. The current state of the contract (https://github.com/kubernetes-sigs/cluster-api/pull/4949/files) specified the following 3 optionally supported "features":
* version
* replicas
* Machines

I think the intention is now (afaik):
* we only support ControlPlanes which support version
* depending on the ControlPlane, we optionally support replicas and machines

Reply:
We can probably make the wording better but the current statement reads says that the `version` in `spec` is needed.
> current contract needs to be expanded to include the support for `version` in the `spec` fields.

AFAIK, when it comes to machine, this information is captured (optionally) in `ClusterClass.ControlPlaneClass` and we do not capture any such information under `cluster.spec.tolopology.controlPlane` - isnt it?

Reply:
It currently states that the contract should be expanded. Imho this is the PR which expands the contract: https://github.com/kubernetes-sigs/cluster-api/pull/4949/files

And I think when #4949 has been merged. The current version of the ClusterClass proposal on this PR does not really make clear that ClusterClass only supports ControlPlane provider which support version. (But I might be totally wrong about all that). Let's wait for feedback from @fabriziopandini or @vincepri .

Reply:
Is it ok if I'm going to open a separated issue/PR to clarify this point as well as the conventions that providers have to follow when developing new templates (so we can keep track on comments on the various PRs)

Reply:
:+1:

Reply:
https://github.com/kubernetes-sigs/cluster-api/issues/5018